app/src/main/java/com/mapbox/maps/testapp/examples/viewport/ViewportShowcaseActivity.kt

```diff
@@ -0,0 +1,243 @@
+      40.0 * pixelDensity,
+      120.0 * pixelDensity,
+      40.0 * pixelDensity
+    )
+  }
+  private val landscapeOverviewPadding: EdgeInsets by lazy {
+    EdgeInsets(
+      30.0 * pixelDensity,
+      380.0 * pixelDensity,
+      20.0 * pixelDensity,
+      20.0 * pixelDensity
+    )
+  }
+  private val followingPadding: EdgeInsets by lazy {
+    EdgeInsets(
+      180.0 * pixelDensity,
+      40.0 * pixelDensity,
+      150.0 * pixelDensity,
+      40.0 * pixelDensity
+    )
+  }
+  private val landscapeFollowingPadding: EdgeInsets by lazy {
+    EdgeInsets(
+      30.0 * pixelDensity,
+      380.0 * pixelDensity,
+      110.0 * pixelDensity,
+      40.0 * pixelDensity
+    )
+  }
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    val binding = ActivityViewportAnimationBinding.inflate(layoutInflater)
+    setContentView(binding.root)
+    viewportButton = binding.switchButton
+    mapView = binding.mapView
+    mapboxMap = binding.mapView.getMapboxMap().apply {
+      addOnStyleLoadedListener {
+        setCamera(
+          CameraOptions.Builder()
+            .zoom(14.0)
+            .center(Point.fromLngLat(POINT_LNG, POINT_LAT))
+            .build()
+        )
+        binding.mapView.location.apply {
+          val provider = FakeLocationProvider()
+          latestLocation = Location(provider.javaClass.simpleName)
+          setLocationProvider(provider)
+          updateSettings {
+            locationPuck = LocationPuck2D(
+              bearingImage = AppCompatResources.getDrawable(
+                this@ViewportShowcaseActivity,
+                R.drawable.mapbox_mylocation_icon_bearing,
+              )
+            )
+          }
+          addOnIndicatorPositionChangedListener {
+            latestLocation.latitude = it.latitude()
+            latestLocation.longitude = it.longitude()
+            viewportDataSource.onLocationChanged(latestLocation)
+            viewportCamera.resetFrame()
+          }
+          addOnIndicatorBearingChangedListener {
+            latestLocation.bearing = it.toFloat()
+            viewportDataSource.onLocationChanged(latestLocation)
+            viewportCamera.resetFrame()
+          }
+        }
+      }
+    }
+    setupViewportCamera()
+  }
+
+  @SuppressLint("SetTextI18n")
+  private fun setupViewportCamera() {
+    viewportCamera = mapView.viewport
+    viewportDataSource = MapboxViewportDataSource(mapboxMap, mapboxMap)
+    viewportCamera.viewportDataSource = viewportDataSource
+    viewportCamera.registerViewportCameraStateChangedObserver { cameraState ->
+      // change title of viewport button depending on the camera state
+      when (cameraState) {
+        ViewportCameraState.TransitionToFollowing,
+        ViewportCameraState.Following -> viewportButton.text = OVERVIEW
+
+        ViewportCameraState.TransitionToOverview,
+        ViewportCameraState.Overview,
+        ViewportCameraState.Idle -> viewportButton.text = FOLLOW
```

Comment:
I'm wondering whether having these states/modes be hard-coded makes sense. Could we instead come up with a more data-driven approach where we expose API that allows developers to define the rules for a given state and then they can add any number of those states to the viewport camera? Transitions between states could have a default implementation, but could also be modeled explicitly if developers want to fine-tune them.

Reply:
To make it easy to use, we could provide some pre-defined configurations that correspond to following, overview, and idle.